File: cpp-implementation/src/others/Stock.cpp
#include "Others.ih"

// Adds a new product to the stock.
void Stock::addProduct(unique_ptr<Product> product) {
    int id = product->getId(); // Get ID of the product

    // Check if a product with this ID already exists in the stock
    if (products.find(id) != products.end()) {
        cerr << "Cannot order product. Product with ID " << id << " is already in use.\n\n";
    } else {
        products[id] = move(product); // Add the product to the stock
        cout << "Product with ID " << id << " ordered from supplier.\n";
    }
}

// Updates an existing product in the stock
void Stock::updateProduct(int id, unique_ptr<Product> product) {
    auto it = products.find(id); // Find the product by its ID

    if (it != products.end()) {
        products[id] = move(product); // Replace the existing product with the new one
        cout << "Product with ID " << id << " updated.\n\n";
    } else {
        cerr << "Cannot update. Product with ID " << id << " not found.\n\n";
    }
}

// Retrieves a product by its ID. Returns a pointer to the product or nullptr if not found.
Product* Stock::getProduct(int id) const {
    auto it = products.find(id); // Find the product by its ID

    if (it != products.end()) {
        return it->second.get(); // Return the product if found
    } else {
        cerr << "Product with ID " << id << " not found.\n\n";
        return nullptr;
    }
}

// Deletes a product from the stock by its ID.
void Stock::deleteProduct(int id) {
    if (products.erase(id) == 0) { // Attempt to erase the product
        cerr << "Cannot delete. Product with ID " << id << " not found.\n\n";
    } else {
        cout << "Product with ID " << id << " deleted.\n\n";
    }
}

std::unique_ptr<Product> Stock::retrieveAndRemoveProduct(int id) {
    auto it = products.find(id);
    if (it != products.end()) {
        std::unique_ptr<Product> product = std::move(it->second); // Move the product out of the map
        products.erase(it); // Erase the product from the map
        return product; // Return the moved product
    } else {
        std::cerr << "Product with ID " << id << " not found in stock.\n";
        return nullptr;
    }
}

File: java-implementation/src/main/java/com/inventory/others/Stock.java
package com.inventory.others;

import java.util.HashMap;
import java.util.Map;
import com.inventory.products.Product;

// Represents a stock of products
public class Stock {
    private Map<Integer, Product> products;

    public Stock() {
        products = new HashMap<>();
    }

    // Adds a new product to the stock.
    public void addProduct(Product product) {
        int id = product.getId(); // Get ID of the product

        // Check if a product with this ID already exists in the stock
        if (products.containsKey(id)) {
            System.err.println("Cannot add product. Product with ID " + id + " is already in use.\n");
        } else {
            products.put(id, product); // Add the product to the stock
            System.out.println("Product with ID " + id + " added to stock.\n");
        }
    }

    // Updates an existing product in the stock
    public void updateProduct(int id, Product product) {
        if (products.containsKey(id)) {
            products.put(id, product); // Replace the existing product with the new one
            System.out.println("Product with ID " + id + " updated.\n");
        } else {
            System.err.println("Cannot update. Product with ID " + id + " not found.\n");
        }
    }

    // Retrieves a product by its ID. Returns the product or null if not found.
    public Product getProduct(int id) {
        if (products.containsKey(id)) {
            return products.get(id); // Return the product if found
        } else {
            System.err.println("Product with ID " + id + " not found.\n");
            return null;
        }
    }

    // Deletes a product from the stock by its ID.
    public void deleteProduct(int id) {
        if (products.remove(id) == null) { // Attempt to remove the product
            System.err.println("Cannot delete. Product with ID " + id + " not found.\n");
        } else {
            System.out.println("Product with ID " + id + " deleted.\n");
        }
    }

    // Retrieves and removes a product from the stock by its ID.
    public Product retrieveAndRemoveProduct(int productId) {
        return products.remove(productId); // This removes and returns the product
    }

    // Allows all the products in stock to be displayed
    public void displayAllProducts() {
        if (products.isEmpty()) {
            System.out.println("The stock is currently empty.");
        } else {
            System.out.println("Current Stock:\n");
            for (Product product : products.values()) {
                System.out.println(product.displayProductInfo());
            }
        }
    }

    // Getter for the products map (if needed)
    public Map<Integer, Product> getProducts() {
        return products;
    }
}
